static/app/views/ddm/useGetChartPalette.tsx

```diff
@@ -0,0 +1,62 @@
+import {useCallback, useRef} from 'react';
+
+import theme from 'sentry/utils/theme';
+
+const CACHE_SIZE = 20; // number of palettes to cache
+
+// We do length - 2 to be aligned with the colors in other parts of the app (copy-pasta)
+// We use Math.max to avoid numbers < -1 as then `getColorPalette` returns undefined (not typesafe because of array access)
+export function getChartColors(seriesCount: number) {
+  return theme.charts.getColorPalette(Math.max(seriesCount - 2, -1));
+}
+
+export function getCachedChartPalette(
+  cache: Record<string, string>[],
+  seriesNames: string[]
+): (seriesName: string) => string {
+  // We do length - 2 to be aligned with the colors in other parts of the app (copy-pasta)
+  // We use Math.max to avoid numbers < -1 as then `getColorPalette` returns undefined (not typesafe because of array access)
+  const defaultColors = getChartColors(seriesNames.length);
+
+  // Check if we already have a palette that includes all of the given seriesNames
+  // We reverse the cache to get the most recent palettes first
+  const cachedPalette = [...cache]
+    .reverse()
+    .find(palette => seriesNames.every(seriesName => seriesName in palette));
+
+  if (cachedPalette) {
+    // Shift the palette to the end of the cache, so we only remove the oldest palette when hitting the cache limit
+    cache.shift();
+    cache.push(cachedPalette);
+    return seriesName => cachedPalette[seriesName] ?? defaultColors[0];
+  }
+
+  // Create a new palette with the default colors ans append at the end of the cache
+  const newPalette = seriesNames.reduce(
+    (palette, seriesName, i) => {
+      palette[seriesName] = defaultColors[i % defaultColors.length];
+      return palette;
+    },
+    {} as Record<string, string>
+  );
+
+  // Single series palettes will always be the same, so we do not need to cache them
+  if (seriesNames.length > 1) {
+    cache.push(newPalette);
+  }
+
+  // Don't cache more than CACHE_SIZE palettes, so we do not create a memory leak
+  if (cache.length > CACHE_SIZE) {
```

Comment:
nit: this means you store cache_size + 1 items at some point.

Reply:
Only in between the last push and shift which is for a few milliseconds, right?

Reply:
Yup. But it implies cache.length > cache.size will only ever be max(size+1, max_cache_size), which if you change the code above can easily be invalidated. If you ever insert more than 1 element, your cache size can grow out of control as you will only ever remove a single cache entry past the size... It would have been safer to use a `while(cache.length > cache.size) ` and enforce than any elements past the cache size are removed

Reply:
k, now I got your point 👍